programs/mango-v4/src/health/cache.rs

```diff
@@ -351,35 +441,135 @@ pub struct HealthCache {
 
-    /// Sum of only the positive health components (assets) and
-    /// sum of absolute values of all negative health components (liabs, always >= 0)
-    pub fn health_assets_and_liabs(&self, health_type: HealthType) -> (I80F48, I80F48) {
-        let mut assets = I80F48::ZERO;
-        let mut liabs = I80F48::ZERO;
-        let sum = |contrib| {
-            if contrib > 0 {
-                assets += contrib;
+    pub fn health_assets_and_liabs_stable_assets(
+        &self,
+        health_type: HealthType,
+    ) -> (I80F48, I80F48) {
+        self.health_assets_and_liabs(health_type, true)
+    }
+    pub fn health_assets_and_liabs_stable_liabs(
+        &self,
+        health_type: HealthType,
+    ) -> (I80F48, I80F48) {
+        self.health_assets_and_liabs(health_type, false)
+    }
+
+    /// Loop over the token, perp, serum contributions and add up all positive values into `assets`
+    /// and (the abs) of negative values separately into `liabs`. Return (assets, liabs).
+    ///
+    /// Due to the way token and perp positions sum before being weighted, there's some flexibility
+    /// in how the sum is split up. It can either be split up such that the amount of liabs stays
+    /// constant when assets change, or the other way around.
+    ///
+    /// For example, if assets are held stable: An account with $10 in SOL and -$12 hupnl in a
+    /// SOL-settled perp market would have:
+    /// - assets: $10 * SOL_asset_weight
+    /// - liabs: $10 * SOL_asset_weight + $2 * SOL_liab_weight
+    /// because some of the liabs are weighted lower as they are just compensating the assets.
+    ///
+    /// Same example if liabs are held stable:
+    /// - liabs: $12 * SOL_liab_weight
+    /// - assets: $10 * SOL_liab_weight
+    ///
+    /// The value `assets - liabs` is the health and the same in both cases.
+    fn health_assets_and_liabs(
+        &self,
+        health_type: HealthType,
+        stable_assets: bool,
+    ) -> (I80F48, I80F48) {
+        let mut total_assets = I80F48::ZERO;
+        let mut total_liabs = I80F48::ZERO;
+        let add = |assets: &mut I80F48, liabs: &mut I80F48, value: I80F48| {
+            if value > 0 {
+                *assets += value;
             } else {
-                liabs -= contrib;
+                *liabs += -value;
             }
         };
-        self.health_sum(health_type, sum);
-        (assets, liabs)
+
+        for token_info in self.token_infos.iter() {
+            // For each token, health only considers the effective token position. But for
+            // this function we want to distinguish the contribution from token deposits from
+            // contributions by perp markets.
+            // However, the overall weight is determined by the sum, so first collect all
+            // assets parts and all liab parts and then determine the actual values.
+            let mut asset_balance = I80F48::ZERO;
+            let mut liab_balance = I80F48::ZERO;
+
+            add(
```

Comment:
508-519, what is the difference between adding balance_spot + hupnl vs simply using balance_spot_and_perp

Reply:
It's a key difference! In this code we allocate the token balance and the perp hupnl separately to assets and liabs. If we only allocated the sum, the result would be way different. Imagine having $5 spot and $-5 SOL-PERP hupnl. This should be accounted as $5 in assets and -$5 in liabs (health ratio 50%), not as 0 (health ratio inf).